presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,30 +66,196 @@
+     * under the read lock.
+     *
+     * The locking hierarchy is as follows:
+     * 1) read/write lock on the cache in PrestoDriver
+     * 2) <internal locking of com.google.common.cache.LoadingCache imlementation>
+     *
+     * With that in mind, the possible scenarios for multiple access are as follows:
+     *
+     * 1) Multiple simultaneous get() operations happening, no close() operation.
+     *    a) No suitable cache entry:
+     *       All callers of get() obtain the read lock
+     *       The cache creates one new entry under its own locking and returns
+     *         it to all callers of get() in some arbitrary sequence.
+     *       All callers atomically increment the refcount on the queryExecutor
+     *    b) Suitable cache entry:
+     *       All callers of get() obtain the read lock
+     *       The cache returns the existing QueryExecutor to all callers of
+     *         get() in some arbitrary sequence.
+     *       All callers atomically increment the refcount on the queryExecutor
+     *
+     * 2) Simultaneous close() operations happening, no get() operation.
+     *    Each thread calling close() gets the write lock in turn and:
+     *       decrements the refcount
+     *       checks to see if the refcount is zero.
+     *       if so, invalidates the cache entry for the queryExecutor
+     *          The cache removes the queryExecutor under its own lock
+     *          and releases its own lock.
+     *       The RemovalListener is called, which closes the HTTP client
+     *       releases the write lock
+     *
+     * 3) Simultaneous get() and close(), get obtains read lock first
+     *    Caller of get() obtains read lock and:
+     *       calls cache.get() which returns existing queryExecutor under the
+     *             cache's own lock
+     *       increments queryExecutor.refcount under readlock
+     *       releases read lock
+     *    Caller of close() obtains write lock and:
+     *       decrements refcount
+     *       refcount != 0
+     *       releases write lock
+     *
+     * 4) Simultaneous close() and get(), close() obtains write lock first
+     *    Caller of close() obtains write lock and:
+     *       decrements refcount
+     *       refcount == 0
+     *       invalidates cache, as in 2 above, closing HTTP client.
+     *       releases write lock
+     *    Caller of get() obtains read lock and:
+     *       calls cache.get() which returns a new queryExecutor under the
+     *             cache's own lock
+     *       increments queryExecutor.refcount under readlock
+     *       releases read lock
+     *
+     * The last case of interest is the process that is using PrestoDriver
+     * calling prestoDriver.close(). This invalidates all of the cached
+     * QueryExecutors regardless of their refcount, and in doing so closes
+     * their HTTP clients. If anybody is still trying to uses a
+     * PrestoConnection created by the driver, they're SOL.
+     */
```

Comment:
This description is very low level. Cannot we make it a bit more succinct leaving implementation details aside (locks etc.), leaving just a general idea here maybe?